### PR TITLE
Allow `[@tailcall true]` and `[@tailcall false]`

### PR DESCRIPTION
**Pull request by gasche**
`[@tailcall true]` is equivalent to `[@tailcall]`, and `[@tailcall false]` warns if the call is in tail-call position.

This is build-up work for the upcoming "TRMC Reborn" PR, which will use these attributes to drive some parts of the TRMC transformation.

### CODE REVIEW

**Comment by gasche** [comment]
Remark: a first cleanup commit in this PR, 66aba00f7c08f73cd953b4f6fea50dc1393a4877, removes the `is_native_tail_call` forward-reference, which is dead code since @nojb moved the `-annot` code in #2141 ( 57d329e07b50e6869875471b085541baadcb8376 ). This forward reference was using backend-specific information to conservatively estimate whether a call was really a tail call (spilling arguments on the stack would break tail-callness), which were only ever used in the annotations produced by the `.annot` file, and in particular not used by the `[@tailcall]` machinery.

One could wonder whether we should try harder to give backend-accurate tailcall information; this is (of course ?) unchanged by the present PR. (On the other hand, having code that raises a warning or not depending on the backend configuration is also bad for usability.) If we decide to do something in this area (no hurry!), I feel somewhat more motivated by the idea of trying harder to preserve tailcalls in presence of spilling, although that may also not be workable at all.

**Comment by gasche** [comment]
This PR is a dependency of the new TRMC PR #9760, and it is short and easy to review independently. Having it merged would help reduce the diff of TRMC, and made that (larger) PR easier to review. Volunteers welcome!

File: utils/warnings.ml
Comment (nojb) [inline utils/warnings.ml]: A better name for this constructor could be `Expected_tailcall`, and it would match up with the other `*-ed_xxx` warnings.

File: lambda/lambda.ml
Comment (nojb) [inline lambda/lambda.ml]: Perhaps these two constructors could be merged into a single one `Expect_tailcall of bool`? It would match up with the warning constructor and the call to `maybe_warn` below...

**Review by nojb — APPROVED** [review]
LGTM. I left a couple of suggestions but feel free to ignore them. One thing that may be nice is to add a test for the "illegal payload" case.

File: utils/warnings.ml
Comment (gasche) [reply utils/warnings.ml]: I think your PR already renamed this constructor to `Tailcall_expected`. Maybe I should change it to `Tailcall_expectation`, now that there are both an "expected" and "*not* expected" case?

Comment (nojb) [reply utils/warnings.ml]: Oops, you are right :) ! `Tailcall_expected of bool` seems fine... But if we are changing it, I find that `Expected_tailcall` reads a bit better.

**Comment by xavierleroy** [comment]
This PR looks almost ready.  What should we do with it?

**Comment by xavierleroy** [comment]
Ping.

File: lambda/lambda.ml
Comment (gasche) [reply lambda/lambda.ml]: Initially I didn't want to do this because I think that `bool` is not clear enough on what it means. (Does `false` mean that the implementation should try to disable tail-call optimization when we are in tailcall position?) But with `Expect` in the name of some form, it is sort of clear that the attribute semantics is (currently) only to report an expectation, so it is in fact equivalent to the current names, but more factorized. Will change.

File: utils/warnings.ml
Comment (gasche) [reply utils/warnings.ml]: I tried with `expected_tailcall` and now warnings on `@tailcall false` read very confusing:

```
Line 3, characters 9-56:
3 |   | n -> (fact_tail [@tailcall false]) (n * acc) (n - 1)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Warning 51 [expected-tailcall]: expected non-tailcall
```

I will try again with `Tailcall_expectation`, which works better with both `true` and `false`.

**Comment by gasche** [comment]
I just rebased it to take into account the comments by Nicolás; this could be good to go if the CI is happy, but I'm in the middle of a deadline (I pushed this to free my "trunk" repository for some other testing) and I'd like to double-check it before merging.

**Comment by gasche** [comment]
I checked the PR and did a last commit, adding a testcase for the "invalid payload" case as suggested by @nojb. This should now be good to go if the CI is happy.

**Comment by nojb** [comment]
The dependencies need to be updated.

**Pull request merged**
